finetune-peft/inference.py

```diff
@@ -0,0 +1,100 @@
+because Maps are not long-term persistent and the data will eventually be lost.
+'''
+
+from beam import Image, endpoint, env, Volume, QueueDepthAutoscaler, experimental
+
+# Path to cache model weights
+VOLUME_PATH = "./gemma-ft"
+FINETUNE_PATH = "./gemma-ft/gemma-2b-finetuned"
+MODEL_PATH = "./gemma-ft/weights"
+
+# This ensures that these packages are only loaded when the script is running remotely on Beam
+if env.is_remote():
+    from transformers import AutoTokenizer, AutoModelForCausalLM
+    from peft import PeftModel
+
+
+def load_finetuned_model():
+    global model, tokenizer, stop_token_ids
+    print("Loading latest...")
+
+    model = AutoModelForCausalLM.from_pretrained(
+        MODEL_PATH, attn_implementation="eager", device_map="auto", is_decoder=True
+    )
+
+    tokenizer = AutoTokenizer.from_pretrained(MODEL_PATH)
+
+    # using our LORA result via the PEFT library
+    model = PeftModel.from_pretrained(model, FINETUNE_PATH)
+    print(model.config)
+
+    stop_token = "<|im_end|>"
+    stop_token_ids = tokenizer.encode(stop_token, add_special_tokens=False)
+
+    s.clear()  # Clear the signal so it doesn't fire again
+
+
+# we are using the experimental Signal abstraction, which lets us hot reload the latest weights
+# without having to restart the inference service. It can be triggered with: experimental.Signal(name="reload-model")
+s = experimental.Signal(
+    name="reload-model",
+    handler=load_finetuned_model,
+)
+
+
+@endpoint(
+    name="gemma-inference",
+    on_start=load_finetuned_model,
+    volumes=[Volume(name="gemma-ft", mount_path=VOLUME_PATH)],
+    cpu=1,
+    memory="16Gi",
+    gpu="T4",
+    image=Image(
+        python_version="python3.9",
+        python_packages=["transformers==4.42.0", "torch", "peft"],
+    ),
+    # this autoscaler will spawn new containers (up to 5) if the queue depth for tasks exceeds 1
+    autoscaler=QueueDepthAutoscaler(max_containers=5, tasks_per_container=1),
+)
+def predict(**inputs):
+    global model, tokenizer, stop_token_ids  # These will have the latest values
+
+    prompt = inputs.get("prompt", None)
+    if not prompt:
+        return {"error": "Please provide a prompt."}
+
+    # now we will format the user provided prompt so that it is of the format that
+    # the fine tuning dataset expects
+    prompt = f"<|im_start|>user\n{prompt}\n<|im_end|>\n<|im_start|>assistant\n"
+
+    # we set the end of sequence token to the last token from <|im_end|>
+    # could probably use stopping criteria to check the full stop matches if  you want to.
+    inputs = tokenizer.encode(prompt, return_tensors="pt").to(model.device)
+    output = model.generate(
+        inputs,
+        max_length=100,
+        num_return_sequences=1,
+        use_cache=False,
+        eos_token_id=stop_token_ids[-1],
+        pad_token_id=tokenizer.eos_token_id,
+    )
+    # here we are trimming the input length from the output so that only the newly generated text is returned
```

Comment:
Nitpick, but the comments should be capitalized (following the conventions here: https://www.cs.cornell.edu/courses/cs1110/2022fa/materials/style/#comment-conventions)